metaci/testresults/models.py

```diff
@@ -274,3 +255,300 @@ class TestResultAsset(models.Model):
+        "cpu_usage_high": FieldType("CPU Usage: High", NearMax("test_cpu_time_used")),
+        "count": FieldType("Count", Count("id")),
+        "failures": FieldType("Failures", Count("id", filter=Q(outcome="Fail"))),
+        "assertion_failures": FieldType(
+            "Assertion Failures",
+            Count("id", filter=Q(message__startswith="System.AssertException")),
+        ),
+        "DML_failures": FieldType(
+            "DML Failures",
+            Count("id", filter=Q(message__startswith="System.DmlException")),
+        ),
+        "other_failures": FieldType(
+            "Other Failures",
+            Count(
+                "id",
+                filter=Q(outcome="Fail")
+                & ~Q(message__startswith="System.DmlException")
+                & ~Q(message__startswith="System.AssertException"),
+            ),
+        ),
+        "success_percentage": FieldType(
+            "Success Percentage",
+            Cast(Count("id", filter=Q(outcome="Pass")), FloatField())
+            / Cast(Count("id"), FloatField())
+            * 100,
+        ),
+    }
+
+    @classmethod
+    def metrics(cls):
+        """Summarize the metrics to {name: calculation} (no displayName)"""
+        return {name: f.aggregation for (name, f) in cls.metric_definitions.items()}
+
+    @classmethod
+    def _get_queryset_for_dates(cls, start_date, end_date, **values):
+        """Get a queryset for a range of dates."""
+        metrics = cls.metrics()
+
+        buildflows = build_models.BuildFlow.objects.filter(
+            time_end__date__gte=start_date, time_end__date__lte=end_date
+        )
+
+        method_contexts = (
+            TestResult.objects.filter(
+                build_flow_id__in=buildflows,
+                build_flow__build__planrepo_id__isnull=False,
+            )
+            .values(
+                "method_id",
+                rel_repo_id=F("build_flow__build__repo"),
+                rel_branch_id=F("build_flow__build__branch"),
+                rel_plan_id=F("build_flow__build__plan"),
+                rel_planrepo_id=F("build_flow__build__planrepo"),
+                **values,
+            )
+            .annotate(
+                agg_count=Count("method__name"),
+                agg_duration_average=metrics["duration_average"],
+                agg_duration_slow=metrics["duration_slow"],
+                agg_duration_fast=metrics["duration_fast"],
+                agg_cpu_usage_average=metrics["cpu_usage_average"],
+                agg_cpu_usage_low=metrics["cpu_usage_low"],
+                agg_cpu_usage_high=metrics["cpu_usage_high"],
+                agg_failures=metrics["failures"],
+                agg_assertion_failures=metrics["assertion_failures"],
+                agg_DML_failures=metrics["DML_failures"],
+                agg_other_failures=metrics["other_failures"],
+            )
+        )
+        return method_contexts
+
+
+class TestResultPerfWeeklySummaryQuerySet(models.QuerySet):
```

Comment:
Weekly? Does this do the same thing as what is already in build.models?

Reply:
If this question is still open, please clarify it.